backend/routers/transcribe.py

```diff
@@ -71,58 +77,67 @@ async def _websocket_util(
-                    if len(audio_buffer) < (frame_size * frames_count):
-                        continue
-
-                    latest_speech_state = get_speech_state(
-                        audio_buffer[:window_size_samples * 10], vad_iterator, window_size_samples
-                    )
-                    if latest_speech_state:
-                        speech_state = latest_speech_state
-
-                    if (voice_found or not_voice) and (voice_found + not_voice) % 100 == 0:
-                        print(uid, '\t', str(int((voice_found / (voice_found + not_voice)) * 100)) + '% \thas voice.')
-
-                    if speech_state == SpeechState.no_speech:
-                        not_voice += 1
-                        # audio_buffer = bytearray()
-                        # continue
+                if codec == 'opus':
+                    decoded_opus = decoder.decode(data, frame_size=320)
+                    samples = torch.frombuffer(decoded_opus, dtype=torch.int16).float() / 32768.0
+                elif codec in ['pcm8', 'pcm16']:  # Both are 16 bit
+                    writable_data = bytearray(data)
+                    samples = torch.frombuffer(writable_data, dtype=torch.int16).float() / 32768.0
+                else:
+                    raise ValueError(f"Unsupported codec: {codec}")
+                
+                audio_buffer.extend(samples)
+                if len(audio_buffer) >= window_size_samples:
+                    tensor_audio = torch.tensor(list(audio_buffer))
+                    if is_speech_present(tensor_audio[-window_size_samples:], vad_iterator, window_size_samples):
+                        if not is_speech_active:
+                            print('+Detected speech')
+                        is_speech_active = True
+                        last_speech_time = time.time()
+                    elif is_speech_active:
+                        if time.time() - last_speech_time > speech_timeout:
+                            is_speech_active = False
+                            # Clear only happens after the speech timeout
+                            audio_buffer.clear()
+                            print('-NO Detected speech')
+                            continue
                     else:
-                        # audio_file.write(audio_buffer.hex() + "\n")
-                        voice_found += 1
-
+                        continue
+            
                 elapsed_seconds = time.time() - timer_start
                 if elapsed_seconds > duration or not socket2:
-                    socket1.send(audio_buffer)
+                    databuffer.extend(data)
+                    if len(databuffer) >= chunk_size:
+                        socket1.send(databuffer[:len(databuffer) - len(databuffer) % chunk_size])
+                        databuffer = databuffer[len(databuffer) - len(databuffer) % chunk_size:]
+                        await asyncio.sleep(REALTIME_RESOLUTION)
```

Comment:
Sleep here is not perfect and can sleep for more or less than REALTIME_RESOLUTION and can cause significant drift over time especially with a low value like 10ms (100x per second)

Here is an example how you could solve offset the issue

https://github.com/deepgram/median-streaming-latency/blob/main/latency.py#L78-L91

Reply:
Yeah for low spec server that surely would cost us the drift. I'll fix that part, and thank you for the reference code! :)